### ISSUE TITLE
[Plugin] OpenTelemetry Related Docker Compose Fails In Multiple Examples

### ISSUE THREAD

**Issue by canteversleep** [Blueprint-uServices/blueprint#176]
We thank the authors for publishing and maintaining this work. We have been trying to run experiments with Blueprint for a paper we are currently writing, but multiple of the example builds have been throwing a similar error in the `docker compose up` step, which has to do with the open. We note that this only occurs in the examples in which the wiring spec uses the open telemetry plugin. We suspect that this is where the issue arises.

We’ve attached the following stack traces where this issue arises:

* DSB_Hotel:
```
=> ERROR [recomd_service_container recomd_service_process 6/6] RUN go build -o /recomd_service_proce  67.3s
 => CANCELED [search_service_container search_service_process 6/6] RUN go build -o /search_service_pr  69.2s
 => [frontend_service_container frontend_service_process 6/6] RUN go build -o /frontend_service_proce  68.4s
 => => # alue of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in variable declaration: *met
 => => # er does not implement "go.opentelemetry.io/otel/metric".Meter (missing method Float64Gauge)
 => => # /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/provider.go:90:9: cannot use mp.meters.Look
 => => # up(s, func() *meter {…}) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in
 => => #  return statement: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing metho
 => => # d Float64Gauge)
 => [rate_service_container rate_service_process 6/6] RUN go build -o /rate_service_process ./rate_se  66.3s
 => => # alue of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in variable declaration: *met
 => => # er does not implement "go.opentelemetry.io/otel/metric".Meter (missing method Float64Gauge)
 => => # /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/provider.go:90:9: cannot use mp.meters.Look
 => => # up(s, func() *meter {…}) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in
 => => #  return statement: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing metho
 => => # d Float64Gauge)
 => [reserv_service_container reserv_service_process 6/6] RUN go build -o /reserv_service_process ./r  66.2s
 => CANCELED [user_service_container user_service_process 6/6] RUN go build -o /user_service_process   62.1s
------
 > [recomd_service_container recomd_service_process 6/6] RUN go build -o /recomd_service_process ./recomd_service_process:
57.62 # go.opentelemetry.io/otel/sdk/metric
57.62 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/meter.go:65:22: cannot use (*meter)(nil) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in variable declaration: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing method Float64Gauge)
57.62 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/provider.go:90:9: cannot use mp.meters.Lookup(s, func() *meter {…}) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in return statement: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing method Float64Gauge)
------
failed to solve: process "/bin/sh -c go build -o /recomd_service_process ./recomd_service_process" did not complete successfully: exit code: 1
```

* SockShop:
```
=> ERROR [payment_ctr payment_proc 6/6] RUN go build -o /payment_proc ./payment_proc                                                                                                                                                                                            61.8s
 => CANCELED [frontend_ctr frontend_proc 6/6] RUN go build -o /frontend_proc ./frontend_proc                                                                                                                                                                                     64.7s
 => [shipping_ctr shipping_proc 6/6] RUN go build -o /shipping_proc ./shipping_proc                                                                                                                                                                                              62.8s
 => => # # go.opentelemetry.io/otel/sdk/metric
 => => # /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/meter.go:65:22: cannot use (*meter)(nil) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in variable declaration: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing
 => => # method Float64Gauge)
 => => # /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/provider.go:90:9: cannot use mp.meters.Lookup(s, func() *meter {…}) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in return statement: *meter does not implement "go.opentelemetry.io/otel/
 => => # metric".Meter (missing method Float64Gauge)
 => [cart_ctr cart_proc 6/6] RUN go build -o /cart_proc ./cart_proc                                                                                                                                                                                                              62.2s
 => => # # go.opentelemetry.io/otel/sdk/metric
 => => # /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/meter.go:65:22: cannot use (*meter)(nil) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in variable declaration: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing
 => => # method Float64Gauge)
 => => # /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/provider.go:90:9: cannot use mp.meters.Lookup(s, func() *meter {…}) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in return statement: *meter does not implement "go.opentelemetry.io/otel/
 => => # metric".Meter (missing method Float64Gauge)
 => CANCELED [order_ctr order_proc 6/6] RUN go build -o /order_proc ./order_proc                                                                                                                                                                                                 60.7s
 => CANCELED [user_ctr user_proc 6/6] RUN go build -o /user_proc ./user_proc                                                                                                                                                                                                     59.2s
------
 > [payment_ctr payment_proc 6/6] RUN go build -o /payment_proc ./payment_proc:
54.82 # go.opentelemetry.io/otel/sdk/metric
54.82 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/meter.go:65:22: cannot use (*meter)(nil) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in variable declaration: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing method Float64Gauge)
54.82 /go/pkg/mod/go.opentelemetry.io/otel/sdk/metric@v1.26.0/provider.go:90:9: cannot use mp.meters.Lookup(s, func() *meter {…}) (value of type *meter) as "go.opentelemetry.io/otel/metric".Meter value in return statement: *meter does not implement "go.opentelemetry.io/otel/metric".Meter (missing method Float64Gauge)
------
failed to solve: process "/bin/sh -c go build -o /payment_proc ./payment_proc" did not complete successfully: exit code: 1
```

**Edit:** changing the wire specs for the aforementioned examples and removing the opentelemetry plugin results in working Docker builds. 

Please advise regarding this issue and regarding anything we may do to support its resolution.

**Comment by vaastav** [comment]
I think this issue is something that is resulting from a dependency version mismatch for the opentelemetry.
I'll look into a fix for this.

**Comment by canteversleep** [comment]
Can you speak to the specific mismatch which may be occurring? We may be able to assist via PR if the issue is trivial.

**Comment by vaastav** [comment]
My best guess is that when Blueprint does `go mod tidy` in the generated code, it starts using opentelemetry version 1.27.0 but the Blueprint dependency is on version 1.26.0. 
Float64Gauge is only available in otel version 1.27.0. 

As a quick workaround, on your local Blueprint repo, you can try upgrading Blueprint's otel dependency to 1.27.0 and that should fix it for now. I'll probably be posting this exact fix by tonight (Pacific Time) after running some tests.

Ideally, we should have a mechanism where we have the ability to specify the dependency of a version when generating code and that is a non-trivial fix (and I am looking into it).

**Comment by vaastav** [comment]
Closing this issue as it was fixed in #178

**Comment by alessandrocornacchia** [comment]
If that can help for the future, reading here [https://pkg.go.dev/go.opentelemetry.io/otel/metric
](https://pkg.go.dev/go.opentelemetry.io/otel/metric
) it says new specifications can be added to the API without a major version bump. Could it be that they added the `Float64Gauge` to the API specifications `go.opentelemetry.io/otel/metric` in  v1.26, but they added an API implementation `go.opentelemetry.io/otel/sdk/metric` of it only starting from v1.27? 


Blueprint using v1.26 for the sdk, then the error.

**Comment by vaastav** [comment]
No, I checked their API specs. The name of the method for the Float64Gauge changed from v1.26 to v1.27  which caused the mismatch.

Regardless, the dependency version in the generated systems should be fixed to v1.26 as well so this shouldn't be an issue in the future.
